### PR TITLE
Integrate gardener testframework and integration tests into testmachinery

### PR DESCRIPTION
**Pull request by schrodit**
**What this PR does / why we need it**:
This PR exposes the creation and deletion of a shoot with the test framework as util-commands to be consumed by the Testmachinery.
These util-commands standardize basic test operations and allow shared usage between different integration tests of various components. 

In addition, TestDefinitions for the Testmachinery are added to the `.test-defs`folder where they are picked up and executed by the Testmachinery.

**Special notes for your reviewer**:
Is the `.test-defs` the right path for these util-commands?

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

File: .test-defs/CreateShoot.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  name: create-shoot
+spec:
+  owner: Gardener-Team
+  description: Tests the creation of a shoot.
+
+  activeDeadlineSeconds: 3600
+  labels: []
+
+  command: [bash, -c]
+  args: ["/tm/setup github.com/gardener gardener && go run $GOPATH/src/github.com/gardener/gardener/.test-defs/cmd/create-shoot"]
```
Comment (rfranzke) [inline .test-defs/CreateShoot.yaml]: In another file you write

```yaml
  - >-
    /tm/setup github.com/gardener gardener &&
    go test $GOPATH/src/github.com/gardener/gardener/test/integration/shoots/applications
    -ginkgo.v -ginkgo.progress
    -kubeconfig=$TM_KUBECONFIG_PATH/gardener.config
    -shootName=$SHOOT_NAME
    -shootNamespace=$PROJECT_NAMESPACE
```

-> Better readability. Can you do the same here, please?

File: .test-defs/DeleteShoot.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  name: delete-shoot
+spec:
+  owner: Gardener-Team
+  description: Tests the deletion of a shoot.
+
+  activeDeadlineSeconds: 1800
+  labels: []
+
+  command: [bash, -c]
+  args: ["/tm/setup github.com/gardener gardener && go run $GOPATH/src/github.com/gardener/gardener/.test-defs/cmd/delete-shoot"]
```
Comment (rfranzke) [inline .test-defs/DeleteShoot.yaml]: Same as above

File: .test-defs/cmd/create-shoot/helper.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
+package main
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/helper.go:15]: license header missing

File: .test-defs/cmd/create-shoot/helper.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	"fmt"
+	"sort"
+
+	gardenv1beta1 "github.com/gardener/gardener/pkg/apis/garden/v1beta1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// getSeed returns the first seed ordered by creation date with the provided cloudprofile.
+func getSeed(gardenClusterClient kubernetes.Interface, cloudprofileName string) (*gardenv1beta1.Seed, error) {
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/helper.go]: Function is not needed, Gardener will automatically determine an appropriate seed cluster itself -> remove

File: .test-defs/cmd/create-shoot/helper.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+	sort.Slice(seeds, func(i, j int) bool {
+		return seeds[i].ObjectMeta.CreationTimestamp.Before(&seeds[j].ObjectMeta.CreationTimestamp)
+	})
+
+	return seeds[0], nil
+}
+
+// getDefaultDomain returns the default domain of a garden cluster by searching the garden namespace for the secret with the label "garden.sapcloud.io/role=default-domain".
+func getDefaultDomain(gardenClusterClient kubernetes.Interface) (string, error) {
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/helper.go]: I don't see where this function is called. Remove it? Why is it needed?

File: .test-defs/cmd/create-shoot/main.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		testLogger.Fatalf("Cannot create shoot artifact %s", err.Error())
+	}
+
+	cloudprofile, secretBindingName, err := getCloudprofile(shootGardenerTest.GardenClient, cloudprovider, projectNamespace)
+	if err != nil {
+		testLogger.Fatalf("Cannot find a cloud profile for %s: %s", cloudprovider, err.Error())
+	}
+
+	seedObject, err := getSeed(shootGardenerTest.GardenClient, cloudprofile.Name)
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: See above, not needed

File: .test-defs/cmd/create-shoot/main.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		testLogger.Fatalf("Cannot find a seed for cloud profile %s: %s", cloudprofile.Name, err.Error())
+	}
+	testLogger.Infof("Use seed %s", seedObject.Name)
+
+	shootObject.Name = shootName
+	shootObject.Namespace = projectNamespace
+	shootObject.Spec.Cloud.Profile = cloudprofile.Name
+	shootObject.Spec.Cloud.Seed = &seedObject.Name
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: See above - remove

File: .test-defs/cmd/create-shoot/main.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+	testLogger.Infof("Successfully created shoot %s", shootName)
+
+	shootTestOperations, err := framework.NewGardenTestOperation(context.TODO(), shootGardenerTest.GardenClient, testLogger, shootObject)
+	if err != nil {
+		testLogger.Fatalf("Cannot create shoot %s", shootName)
+	}
+
+	err = shootTestOperations.DownloadKubeconfig(context.TODO(), shootTestOperations.SeedClient, shootTestOperations.ShootSeedNamespace(), gardenv1beta1.GardenerName, fmt.Sprintf("%s/shoot.config", kubeconfigPath))
+	if err != nil {
+		testLogger.Fatalf("Cannot download shoot kubeconfig")
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: log the error?

File: .test-defs/cmd/create-shoot/main.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+	shootGardenerTest.Shoot = shootObject
+	shootObject, err = shootGardenerTest.CreateShoot(context.TODO())
+	if err != nil {
+		testLogger.Fatalf("Cannot create shoot %s: %s", shootName, err.Error())
+	}
+	testLogger.Infof("Successfully created shoot %s", shootName)
+
+	shootTestOperations, err := framework.NewGardenTestOperation(context.TODO(), shootGardenerTest.GardenClient, testLogger, shootObject)
+	if err != nil {
+		testLogger.Fatalf("Cannot create shoot %s", shootName)
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: log the error?

File: .test-defs/cmd/create-shoot/helper.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+		return "", err
+	}
+	if len(secrets.Items) == 0 {
+		return "", fmt.Errorf("Cannot find a default domain")
+	}
+	return secrets.Items[0].Annotations["dns.garden.sapcloud.io/domain"], nil
+}
+
+// getSecretBindingName returns the name of the secretbinding that corresponds to a cloudprofile in a project namespace.
+func getSecretBindingName(gardenClient kubernetes.Interface, projectNamespace, cloudprofileName string) (string, error) {
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/helper.go]: `getSecretBindingName ` and `getCloudprofile` - why is this needed in the first place? Can't we just get these variables from the ENV (like some other inputs)?

> `LabelSelector: fmt.Sprintf("cloudprofile.garden.sapcloud.io/name=%s", cloudprofileName)`

This label is no contract/not part of any API, so you can't rely on it - it's just used by the dashboard.

File: .test-defs/cmd/create-shoot/helper.go [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	if len(secretBindings.Items) == 0 {
+		return "", fmt.Errorf("Cannot find a secretbindings for cloudprofile %s in namespace %s", cloudprofileName, projectNamespace)
+	}
+	return secretBindings.Items[0].Name, nil
+}
+
+// getCloudprofile searches a garden cluster for cloudprofiles of the provided cloudprovider with a secretbinding in the given project namespace.
+// It returns the cloudprofile, the secretBinding name and any error encountered.
+func getCloudprofile(gardenClusterClient kubernetes.Interface, cloudprovider gardenv1beta1.CloudProvider, projectNamespace string) (*gardenv1beta1.CloudProfile, string, error) {
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/helper.go]: See above - better get the to-be-used cloud profile via env.

File: .test-defs/cmd/create-shoot/main.go [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+	if cloudprovider == "" {
+		testLogger.Fatalf("EnvVar 'CLOUDPROVIDER' needs to be specified")
+	}
+	k8sVersion = os.Getenv("K8S_VERSION")
+	if k8sVersion == "" {
+		testLogger.Fatalf("EnvVar 'K8S_VERSION' needs to be specified")
+	}
+
+	machineType = os.Getenv("MACHINE_TYPE")
+	if os.Getenv("AUTOSCALER_MIN") != "" {
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]:
```suggestion
	if autoscalerMin := os.Getenv("AUTOSCALER_MIN"); autoscalerMin != "" {
```

File: .test-defs/cmd/create-shoot/main.go [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+		min, err := strconv.Atoi(os.Getenv("AUTOSCALER_MIN"))
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]:
```suggestion
		min, err := strconv.Atoi(autoscalerMin)
```

File: .test-defs/cmd/create-shoot/main.go [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+		if err != nil {
+			testLogger.Infof("Cannot parse %s: %s", os.Getenv("AUTOSCALER_MIN"), err.Error())
+		}
+		autoScalerMin = &min
+	}
+	if os.Getenv("AUTOSCALER_MAX") != "" {
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: See above suggestion

File: .test-defs/cmd/create-shoot/main.go [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		testLogger.Fatalf("Cannot create ShootGardenerTest %s", err.Error())
+	}
+
+	_, shootObject, err := framework.CreateShootTestArtifacts(fmt.Sprintf("example/90-shoot-%s.yaml", cloudprovider), "")
+	if err != nil {
+		testLogger.Fatalf("Cannot create shoot artifact %s", err.Error())
+	}
+
+	cloudprofile, secretBindingName, err := getCloudprofile(shootGardenerTest.GardenClient, cloudprovider, projectNamespace)
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: See above, hopefully not needed

File: .test-defs/cmd/delete-shoot/main.go [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+	}
+
+	err = shootGardenerTest.DeleteShoot(context.TODO())
+	if err != nil {
+		if errors.IsNotFound(err) {
+			testLogger.Infof("Shoot %s cannot be found", shootName)
+			return
+		}
+		testLogger.Fatalf("Cannot delete shoot %s: %s", shootName, err.Error())
+	}
```
Comment (rfranzke) [inline .test-defs/cmd/delete-shoot/main.go:65]:
```suggestion
	if err := shootGardenerTest.DeleteShoot(context.TODO()); err != nil && !errors.IsNotFound(err) {
		testLogger.Fatalf("Cannot delete shoot %s: %s", shootName, err.Error())
	}
```

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: .test-defs/cmd/create-shoot/helper.go (hunk 10 above)
Comment (schrodit) [reply .test-defs/cmd/create-shoot/helper.go]: Is there another way to get any parameters programmatically (like region) or is it better to get all needed parameters with env vars?
We want the input parameters to be as little as possible.

Comment (rfranzke) [reply .test-defs/cmd/create-shoot/helper.go]: I'd recommend to get these things (cloudprofile, region, secret) as inputs (just like it was with the old integration test).
For cloudprofile/region you could theoretically list all seeds and get one that matches to the desired cloud, however, you would not be able to find a proper secretbinding programmatically (this is not coupled in any way to the cloud provider - as said, the label is just for the dashboard).
--> Better get everything as inputs.

File: .test-defs/CreateShoot.yaml [hunk 17]
```diff
@@ -0,0 +1,17 @@
+kind: TestDefinition
+metadata:
+  name: create-shoot
+spec:
+  owner: Gardener-Team
```
Comment (OlegLoewen) [inline .test-defs/CreateShoot.yaml]: owner should be a valid email. Check also other test-defs

File: .test-defs/cmd/create-shoot/helper.go (hunk 10 above)
Comment (schrodit) [reply .test-defs/cmd/create-shoot/helper.go]: Done

**Review by OlegLoewen — CHANGES_REQUESTED** [review]

File: .test-defs/CreateShoot.yaml (hunk 17 above)
Comment (schrodit) [reply .test-defs/CreateShoot.yaml]: Done

File: .test-defs/cmd/create-shoot/main.go [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+	if zone == "" {
+		testLogger.Fatalf("EnvVar 'ZONE' needs to be specified")
+	}
+	k8sVersion = os.Getenv("K8S_VERSION")
+	if k8sVersion == "" {
+		testLogger.Fatalf("EnvVar 'K8S_VERSION' needs to be specified")
+	}
+
+	machineType = os.Getenv("MACHINE_TYPE")
+	if autoscalerMax := os.Getenv("AUTOSCALER_MIN"); autoscalerMax != "" {
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: confusing variable name (`max` vs. what you read from the env `min`)

File: .test-defs/cmd/create-shoot/main.go [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+		min, err := strconv.Atoi(autoscalerMax)
+		if err != nil {
+			testLogger.Infof("Cannot parse %s: %s", os.Getenv("AUTOSCALER_MIN"), err.Error())
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: use the variable name (`autoscalerMin/Max`) here instead of reading from the env again

File: .test-defs/cmd/create-shoot/main.go [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+		}
+		autoScalerMin = &min
+	}
+	if autoscalerMin := os.Getenv("AUTOSCALER_MAX"); autoscalerMin != "" {
+		max, err := strconv.Atoi(autoscalerMin)
+		if err != nil {
+			testLogger.Infof("Cannot parse %s: %s", os.Getenv("AUTOSCALER_MAX"), err.Error())
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: use the variable name (`autoscalerMin/Max`) here instead of reading from the env again

File: .test-defs/cmd/create-shoot/main.go [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+
+	machineType = os.Getenv("MACHINE_TYPE")
+	if autoscalerMax := os.Getenv("AUTOSCALER_MIN"); autoscalerMax != "" {
+		min, err := strconv.Atoi(autoscalerMax)
+		if err != nil {
+			testLogger.Infof("Cannot parse %s: %s", os.Getenv("AUTOSCALER_MIN"), err.Error())
+		}
+		autoScalerMin = &min
+	}
+	if autoscalerMin := os.Getenv("AUTOSCALER_MAX"); autoscalerMin != "" {
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: confusing variable name (`max` vs. what you read from the env `min`)

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: .test-defs/cmd/create-shoot/main.go [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+	if zone == "" && cloudprovider != gardenv1beta1.CloudProviderAzure {
+		testLogger.Fatalf("EnvVar 'ZONE' needs to be specified")
+	}
+	k8sVersion = os.Getenv("K8S_VERSION")
+	if k8sVersion == "" {
+		testLogger.Fatalf("EnvVar 'K8S_VERSION' needs to be specified")
+	}
+
+	machineType = os.Getenv("MACHINE_TYPE")
+	if autoScalerMaxEnv := os.Getenv("AUTOSCALER_MIN"); autoScalerMaxEnv != "" {
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: I'm still confused. The variable name is `autoscalerMax`, but you read `AUTOSCALER_MIN` from env?!

File: .test-defs/cmd/create-shoot/main.go [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+		autoScalerMaxInt, err := strconv.Atoi(autoScalerMaxEnv)
+		if err != nil {
+			testLogger.Infof("Cannot parse %s: %s", autoScalerMaxEnv, err.Error())
+		}
+		autoScalerMin = &autoScalerMaxInt
+	}
+	if autoScalerMinEnv := os.Getenv("AUTOSCALER_MAX"); autoScalerMinEnv != "" {
```
Comment (rfranzke) [inline .test-defs/cmd/create-shoot/main.go]: I'm still confused. The variable name is `autoscalerMax`, but you read `AUTOSCALER_MIN` from env?!

File: .test-defs/cmd/create-shoot/main.go (hunk 22 above)
Comment (schrodit) [reply .test-defs/cmd/create-shoot/main.go]: Oh sorry, my bet. I mixed up min and max.

**Review by rfranzke — APPROVED** [review]
/lgtm
